lib/srv/forward/sshserver.go

```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
 	return nil
 }
 
+// handleX11ChannelRequest accepts an X11 channel and forwards it back to the client.
+// Servers which support X11 forwarding request a separate channel for serving each
+// inbound connection on the X11 socket of the remote session.
+func (s *Server) handleX11ChannelRequest(ctx context.Context, xreq ssh.NewChannel) error {
+	// accept inbound X11 channel from server
+	sch, sin, err := xreq.Accept()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// setup outbound X11 channel to client
+	cch, cin, err := s.sconn.OpenChannel(sshutils.X11ChannelRequest, xreq.ExtraData())
+	if err != nil {
+		sch.Close()
+		return trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// scw and ccw protect CloseWrite calls against sch and cch respectively.
+	var scw, ccw sync.Once
+
+	// ioflag is used by the two io goroutines to determine if they are the
+	// first or second goroutine to complete.  The second goroutine to complete
+	// ensures that the main request forwarding goroutine also terminates by
+	// cancelling its context.
+	var ioflag uint32
+
+	defer func() {
+		// consume our sync.Once instances if they haven't been
+		// consumed already (prevents CloseWrite calls from racing
+		// with Close calls).
+		scw.Do(func() {})
+		ccw.Do(func() {})
+		cancel()
+		sch.Close()
+		cch.Close()
+	}()
+
+	go func() {
+		// forward data from server to client
+		io.Copy(cch, sch)
+		// inform client that no more data is coming
+		ccw.Do(func() {
+			cch.CloseWrite()
+		})
+		// if we are the second io forwarder to complete,
+		// then signal shutdown to main goroutine.
+		if atomic.SwapUint32(&ioflag, 1) > 0 {
+			cancel()
+		}
+	}()
+
+	go func() {
+		// forward data from client to server
+		io.Copy(sch, cch)
+		// inform server that no more data is coming
+		scw.Do(func() {
+			sch.CloseWrite()
+		})
+		// if we are the second io forwarder to complete,
+		// then signal shutdown to main goroutine.
+		if atomic.SwapUint32(&ioflag, 1) > 0 {
+			cancel()
+		}
+	}()
+
+	// multiplex requests until channel is closed or both
+	// io forwarders have completed.
+	for {
+		select {
+		case sreq := <-sin:
+			if sreq == nil {
+				return nil
```

Comment:
Do we need to return right away on channel closure?
If not, how about "disabling" the select case for a closed channel (setting `sin` to `nil`) in this case?
This will make it so that the function only returns after both `io.Copy` goroutines have finished, which in turn makes those `sync.Once`s unnecessary.